tony-core/src/main/java/com/linkedin/tony/ApplicationMaster.java

```diff
@@ -853,38 +859,53 @@ public String registerWorkerSpec(String taskId, String spec) throws IOException
         killChiefWorkerIfTesting(taskId);
       }
 
-      // Return null until all expected tasks have registered
-      int numExpectedTasks = session.getNumExpectedTasks();
-      if (registeredTasks.size() == numExpectedTasks) {
-        LOG.info("All " + numExpectedTasks + " expected tasks registered.");
-        return getClusterSpec();
-      } else {
-        // Periodically print a list of all tasks we are still awaiting registration from.
-        if (System.currentTimeMillis() - lastRegisterWorkerTime > REGISTRATION_STATUS_INTERVAL_MS) {
-          Set<TonyTask> unregisteredTasks = getUnregisteredTasks();
-          LOG.info(String.format("Received registrations from %d tasks, awaiting registration from %d tasks.",
-              registeredTasks.size(), numExpectedTasks - registeredTasks.size()));
-          unregisteredTasks.forEach(t -> {
-            // Stop application when timeout
-            if (registrationTimeoutMs > 0 && System.currentTimeMillis() - t.getStartTime() > registrationTimeoutMs) {
-              String errorMsg = String.format("Stopping AM for task [%s:%s] registration timeout: "
-                              + "allocated container is %s on host %s",
-                  t.getJobName(), t.getTaskIndex(),
-                  (t.getContainer() != null ? t.getContainer().getId().toString() : "none"),
-                  (t.getContainer() != null ? t.getContainer().getNodeId().getHost() : "none"));
-              LOG.error(errorMsg);
-              session.setFinalStatus(FinalApplicationStatus.FAILED, errorMsg);
-              stop();
-            } else {
-                LOG.info(String.format("Awaiting registration from task %s %s in %s on host %s",
+      // two distributed mode(default is cluster) cases:
+      // 1. when in independent-task mode, task will be allowed to run when AM accept worker registered spec,
+      // 2. when in cluster mode, it will start until all tasks have registered.
+      switch (distributedMode) {
+        case CLUSTER:
```

Comment:
CLUSTER, INDEPENDENT_TASK is not accurate, cluster mode doesn't have to wait either. What about GANG and FCFS modes?

Reply:
Good. FCFS means first come first schedule?

Reply:
Done